src/constructApplications.js

```diff
@@ -0,0 +1,92 @@
+import { resolvePath } from "./matchRoute";
+
+/**
+ * @typedef {{
+ * routes: import('./constructRoutes').ResolvedRoutesConfig;
+ * loadApp: LoadApp;
+ * }} ApplicationOptions
+ *
+ * @typedef {(applicationName: string) => Promise<import('single-spa').Application>} LoadApp
+ *
+ * @typedef {{
+ * app: (config: import('single-spa').AppProps) => Promise<import('single-spa').LifeCycles>
+ * }} WithLoadFunction
+ *
+ * @param {ApplicationOptions} applicationOptions
+ * @returns {Array<import('single-spa').RegisterApplicationConfig & WithLoadFunction>}
+ */
+export function constructApplications({ routes, loadApp }) {
+  const applicationMap = {};
+  recurseRoutes(applicationMap, routes.base, {}, routes.routes);
+
+  /**
+   * @type {Array<{
+   * name: string;
+   * customProps: object;
+   * activeWhen: import('single-spa').Activity
+   * }>}
+   */
+  const partialApplications = Object.keys(applicationMap).map((name) => {
+    const resolvedRoutes = applicationMap[name];
+    return {
+      name,
+      customProps: (_name, location) => {
```

Comment:
This relies on https://github.com/single-spa/single-spa/pull/533 which was released in single-spa@5.4.0